### PR TITLE
public_key: Fix hotsname comparison bug

### LINKED ISSUE #8021: Unexpected matching rule of `.` appears in the hostname verification func
**Describe the bug**
The prefix `.` (0x2e) is ignored in matching `CommonName (CN)` when using `public_key:pkix_verify_hostname`. E.g., the `[CN]='..a'(0x2e2e61)` can match the `[input]='a'(0x61)`. According to the rules in Stringprep(RFC3454, RFC4518) and PKIX(RFC 5280, RFC 6125), there is no rule about ignoring dot (.) at the prefix of string. It would be a name confusion if two strings are falsely matched in the hostname verification task. So we want to inquire if this is a bug. Looking forward to your views.

**To Reproduce**
Example code for calling `OTP`'s API for hostname verification.
```Erlang
-module(test_dn).
-export([verify/0]).
verify() ->
	{ok, [[EE]]} = init:get_argument(ee),
	{ok, EEPemBin} = file:read_file(EE),
	[EEPemEntries] = public_key:pem_decode(EEPemBin),
	{_, EeBytes, _} = EEPemEntries,
	EeCertOtp = public_key:pkix_decode_cert(EeBytes, otp),

	{ok, [[Name]]} = init:get_argument(n),
	io:format("~p: ", [Name]),
	case public_key:pkix_verify_hostname(EeCertOtp, [{dns_id, Name}]) of
		true -> io:fwrite("ok\n"), halt(0);
		false -> io:fwrite("name checking failed"), halt(1)
	end.
```
Example test case of X.509 certificate. 
```
-----BEGIN CERTIFICATE-----
MIIDOzCCAiOgAwIBAgICECEwDQYJKoZIhvcNAQELBQAwYzELMAkGA1UEBhMCVVMx
ITAfBgNVBAoTGFRoZSBHbyBEYWRkeSBHcm91cCwgSW5jLjExMC8GA1UECxMoR28g
RGFkZHkgQ2xhc3MgMiBDZXJ0aWZpY2F0aW9uIEF1dGhvcml0eTAeFw0xNjAyMDcx
NzI0MDBaFw0yNDEyMTAwNTA3MTRaMA4xDDAKBgNVBAMTAy4uYTCCASAwDQYJKoZI
hvcNAQEBBQADggENADCCAQgCggEBAKt8ORTxpKElswQ8QtFdXPcB642LllZ/wVmD
HDxvxeubnIb/eBMW8gl4FYyeJoXoGV7YWpKgBiC+LO7rCdsGMhzlYKyhx61LgXYD
M57TkFsTmjk1DoBQUuODX0V4kgf8Td2Gv6DETre/0eEMrNkCTDq6W+d6UHhMor/+
rsK986kqNxDFweyc/2BhlF6yXFpl5BFnXEwEOwRTw0Zjx6jABnOLYAfxZu4oGiSW
T0Zb2Z5vbrq0kuDQ5t2zjFvEUj9pxkoR/QgWsKwA6nRPg2L3BdnLG98dvIZqFyEn
tsKWQ3Z6x8TP/4GSyGfbjoVAmYQJQ5DAoo2MCk8kcEqjZwc5T5kCAQOjUDBOMAwG
A1UdEwEB/wQCMAAwHQYDVR0OBBYEFCTNEK+XDJUYJ45yRL+l+zNeru17MB8GA1Ud
IwQYMBaAFJcswpcBl9XDsKD5xI6eZbKBX8qrMA0GCSqGSIb3DQEBCwUAA4IBAQCp
1nnkQrCtd31yT8cPcvnE83pS8eesGSrdblkzjcvgV56SG9+/boi6/U6u0rbGDqmx
HzmVtg6YsAz4km56sgdPBeVAh4Rm3TjzeFuhCw1YjqSrvL3MVZIMNdbY5Rs2SpxT
szPc5J139UWS5VQpZrwoPOqD7ny72ve1QbNZbkERiY2mC5fJ7/pdPMzms4d/+GOf
G+YMoIf9xYIUexKImHI1OfiLG3PcBd04qolgp/p4k0O3DaF8HBsK4IukWuFGhTUB
zWvN5QOVLYdrVEpF4tvrSb6EKzNp3VQ2ucweKGdF7hNs9Wv1KotnxnScolavfBCW
1XWBeXZVmw1bennWAlfr
-----END CERTIFICATE-----
```
Examples of input strings: `a`, `.a`, `..a`, etc.
Example of execution:
```bash
erlc test_dn.erl # compile
erl -noshell -run test_dn verify -ee <certFile> -n <inputStr> -s init stop # run
```

**Expected behavior**
It's the same as the description.

**Affected versions**
Testing environment: Erl 22. Not sure about other versions.

**Additional context**
Note that this behavior appears in matching `CN`. So, the certificate in the test case does not have the SAN part in certificate extensions. The matching CN as hostname rule is in the RFC 6125. 

From the functionality of `CN` in a certificate, there is no rule that the name string and input string should be fully qualified domain names (FQDN). Such strings like `..a` should be acceptable to be stored in an X.509 certificate.

### PR DESCRIPTION
**Pull request by IngelaAndin**
Closes #8021

### CODE REVIEW

File: lib/public_key/test/public_key_SUITE.erl [hunk 1]
```diff
@@ -1132,8 +1132,11 @@ pkix_verify_hostname_options(Config) ->
     true = public_key:pkix_verify_hostname(Cert, [{dns_id,"abb.bar.example.com"}]),
     false = public_key:pkix_verify_hostname(Cert, [{dns_id,"example.com"},
                                                    {dns_id,"abb.bar.example.com"}],
-                                            [{fqdn_fun,fun(_)->undefined end}]).
-    
+                                            [{fqdn_fun,fun(_)->undefined end}]),
+    {ok, PrefixBin} = file:read_file(filename:join(DataDir,"prefix-dots.pem")),
```
Comment (u3s) [inline lib/public_key/test/public_key_SUITE.erl:1136]: can you add a code comment explaining what is specific about prefix-dots.pem  and relevant for test lines below?
I guess maybe this: "Subject: CN = ..a"

File: lib/public_key/test/public_key_SUITE.erl [hunk 2]
```diff
@@ -1132,8 +1132,11 @@ pkix_verify_hostname_options(Config) ->
...
+    PrefixCert = public_key:pkix_decode_cert(element(2,hd(public_key:pem_decode(PrefixBin))), otp),
+    true = public_key:pkix_verify_hostname(PrefixCert, [{dns_id,"..a"}]),
```
Comment (u3s) [inline lib/public_key/test/public_key_SUITE.erl:1138]: is `public_key:pkix_verify_hostname(PrefixCert, [{dns_id,"a"}])` interesting to test as well?
it appeared in the discussion AFAIK ...

**Review by dgud — APPROVED** [review]

File: lib/public_key/test/public_key_SUITE.erl (hunk 1 above)
Comment (IngelaAndin) [reply lib/public_key/test/public_key_SUITE.erl:1136]: Yes subject "CN=..a" and the whole name should match.  ".a" should not match "..a" which it wrongly did!

File: lib/public_key/test/public_key_SUITE.erl (hunk 2 above)
Comment (IngelaAndin) [reply lib/public_key/test/public_key_SUITE.erl:1138]: No the problem was with more than one dot.

**Review by u3s — APPROVED** [review]

File: lib/public_key/test/public_key_SUITE.erl (hunk 1 above)
Comment (IngelaAndin) [reply lib/public_key/test/public_key_SUITE.erl:1138]: Add a comment

**Pull request merged**